[frio] Easy image attach

xundeenergie opened the pull request:
This PR adds functionality to add images via drag&drop and copy&paste to 

- creating a new posting with /compose
-  creating a new posting with jot
- adding a comment

It resolves https://github.com/friendica/friendica/issues/7037

A new package for php is needed: enyo/dropzone, it's added to composer.json.

Attaching an Image via D&D or C&P adds also a preview-thumbnail to the dialog and the url for the image to the textarea, which is one, where you just write your comment or posting.

![grafik](https://user-images.githubusercontent.com/5894829/224599183-cc94d698-26d2-4ecc-af07-03bad012f27f.png)

In jot overlay you can see (how in every other case too), a field "Drop files here to upload". You can touch/click this, and get an upload Dialog. Or you drop/paste Files just somewhere on the dialog.

![grafik](https://user-images.githubusercontent.com/5894829/224599457-c6fa3068-a2e7-4fc2-acd5-91293c441f38.png)

![grafik](https://user-images.githubusercontent.com/5894829/224599547-ea5adfeb-2ba0-4366-bdc5-c373b4f5900a.png)


And creating a comment works also

![grafik](https://user-images.githubusercontent.com/5894829/224599614-35ecb461-cff6-4350-b73c-3dfc00ab8722.png)

Some fixes are outstanding (Error on creating Dropzone, because it extis...), so it's a WIP.

Please try this PR in develop and enjoy

For code formatting-purposes feel free, to give me hints, i will fix them.


For trying it, you have to run `bin/composer.phar update && bin/composer.phar install`

Commit:
fix path for js and css loading

Commit:
fix dropzone class in template

Commit:
add missing css

Commit:
add enyo/dropzone

Commit:
fix composer dropzone

MrPetovan commented on view/templates/item/compose.tpl:
```diff
@@ -94,3 +95,43 @@
 </div>
+<script>
+  Dropzone.autoDiscover = false;
+  var dropzoneCompose = new Dropzone( '#comment-edit-form-{{$id}}',  { 
+    paramName: "userfile", // The name that will be used to transfer the file
+    maxFilesize: 2, // MB
+    previewsContainer: '#dz-previewsCompose',
+    preventDuplicates: true,
+    clickable: true,
+    thumbnailWidth: 100,
+    thumbnailHeight: 100,
+    url: "/media/photo/upload?response=url&album=",
+    accept: function(file, done) {
+      done();
+    },
+    init: function() {
+        this.on("success", function(file, serverResponse) {
+                var target = $('#comment-edit-text-{{$id}}')
+                var resp = $(serverResponse).find('div#content').text()
+                if (target.setRangeText) {
+                    //if setRangeText function is supported by current browser
+                    target.setRangeText(" " + $.trim(resp) + " ")
+                } else {
+                    target.focus()
+                    document.execCommand('insertText', false /*no UI*/, $.trim(resp));
+                }
+        });
+    },
+  });
+
+  $('#comment-edit-form-{{$id}}').on('paste', function(event){
+    const items = (event.clipboardData || event.originalEvent.clipboardData).items;
+    items.forEach((item) => {
+      if (item.kind === 'file') {
+        // adds the file to your dropzone instance
+        console.log(item);
+        dropzoneCompose.addFile(item.getAsFile())
+      }
+    })
+  });
```
Please indent using tabulations.

MrPetovan commented on view/templates/item/compose.tpl:
```diff
@@ -94,3 +95,43 @@
+  Dropzone.autoDiscover = false;
+  var dropzoneCompose = new Dropzone( '#comment-edit-form-{{$id}}',  { 
+    paramName: "userfile", // The name that will be used to transfer the file
+    maxFilesize: 2, // MB
```
This value should come from the internal size limit through a template variable. This size is given by `Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'))` but you have to divide by either 10^6 (MB) or 2^20 (Mb) to get the value for this library.

xundeenergie replied:
Sorry... i tried something, but i have no idea, how to do this.
When i add declaration of variable in php, the composer returns blank white... 
I need help for this.

xundeenergie replied:
I add to `src/Module/Item/Compose.php` the following line
```
                  $tpl = Renderer::getMarkupTemplate('item/compose.tpl');
                  return Renderer::replaceMacros($tpl, [
+                         '$max_imagesize'    => Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')),
                          '$l10n' => [
....
```
and in `view/templates/item/compose.tpl` i added this debug-output:

```
  <script>
    Dropzone.autoDiscover = false;
+   var maxis = {{$max_imagesize}};
+   console.log("maxis", maxis);
    var dropzoneCompose = new Dropzone( '#comment-edit-form-{{$id}}',  { 
```

and then /compose gives me a white, blank page.

When i set a string in Compose.php, then it writes this string in the console... 
When i just set `'$max_imagesize'    => DI::config()->get('system', 'maximagesize'),`
then i get `maxis 800000` in console.

So the "getBytesFromShorthand()" produces an error...

xundeenergie replied:
The error seems in
-  `src/Content/Conversation.php`
- `src/Module/Post/Edit.php` when editing a post/comment`

I testes much around and found out, that only in this file 
`'$max_imagesize'    => DI::config()->get('system', 'maximagesize'),`
does not work.

So i set on the not working stanzas a manual value and the command in commend... 
Maybe you have some ideas?

MrPetovan replied:
Sorry for my partially functional example, you need to add `use Friendica\Util\Strings;` near the top of the `src/Module/Item/Compose.php` file for my snippet to work. Alternatively, you can use the following portable snippet: `\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'))`.

xundeenergie replied:
This is the error only with unmodified code, just on editing with DI:: set.
```
2023-03-13T16:14:43Z index [ERROR]: Uncaught Exception Error: "Class "Friendica\Module\Post\DI" not found" at /var/www/soc.schuerz.at/src/Module/Post/Edit.php line 186 {"exception":"Error: Class \"Friendica\\Module\\Post\\DI\" not found in /var/www/soc.schuerz.at/src/Module/Post/Edit.php:186\nStack trace:\n#0 /var/www/soc.schuerz.at/src/BaseModule.php(244): Friendica\\Module\\Post\\Edit->content()\n#1 /var/www/soc.schuerz.at/src/App.php(700): Friendica\\BaseModule->run()\n#2 /var/www/soc.schuerz.at/index.php(52): Friendica\\App->runFrontend()\n#3 {main}"} - {"file":null,"line":null,"function":null,"request-id":"640f4bf378653","uid":"49715c","process_id":1253003}
```

xundeenergie replied:
ahhhh!!

adding `use Friendica\DI;` on top of the file solves the error. So i will fix the rest as expected now.

MrPetovan commented on view/templates/item/compose.tpl:
```diff
@@ -94,3 +95,43 @@
+    items.forEach((item) => {
+      if (item.kind === 'file') {
+        // adds the file to your dropzone instance
+        console.log(item);
```
Please remove debug output from your code when you're done with the PR.

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,106 @@ function htmlToText(htmlString) {
-		$(".item-" + ident + " .button-event:not(#" + verb + "-" + ident + ")").removeClass("active");
+	_verb = un ? 'un' + verb : verb;
+	var thumbsClass = '';
+	switch (verb) {
+		case 'like':
+			thumbsClass = 'fa-thumbs-up';
+			break;
+		case 'dislike':
+			thumbsClass = 'fa-thumbs-down';
+			break;
+		case 'announce':
+			thumbsClass = 'fa-retweet';
+	}
+	if (verb.indexOf('announce') === 0 ) {
+		// Share-Button(s)
+		// remove share-symbol, to replace it by rotator
+		$('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child').removeClass('fa-share');
+		$('button[id^=announce-' + ident.toString() + '] i:first-child').removeClass('fa-retweet');
+		// if no wait-rotator for activity(verb) is added, add it. or just show it, if exists
+		if ($('img[id^=waitfor-' + verb + '-' + ident.toString() + ']').length == 0) {
+			// append rotator to the shareMenu-button for small media
+      $('<img>')
+        .attr('id', 'waitfor-' + verb + '-' + ident.toString())
+				.addClass('fa')
+				.appendTo($('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child' )).show();
+		} else {
+			$('<img>').attr({id: 'waitfor-' + verb + '-' + ident.toString(), src: 'images/rotator.gif'})
+		}
+	}
+	$('button[id^=' + verb + '-' + ident.toString() + '] i:first-child').removeClass(thumbsClass);
+	// if verb is announce, then one rotator is added above to the shareMedia-dropdown button
+	if ($('button:not(button.dropdown-toggle) img#waitfor-' + verb + '-' + ident.toString()).length == 0) {
+		$('<img>')
+      .attr('id', 'waitfor-' + verb + '-' + ident.toString())
+			.attr('src', 'images/rotator.gif')
+			.addClass('fa')
+			.appendTo($('button[id^=' + verb + '-' + ident.toString() + '] i:first-child')).show();
+	} else {
+		// show existing rotator for activity
+		$('img[id^=waitfor-' + verb + '-' + ident.toString() + ']').show()
```
If the drag & drop image upload is independent from the asynchronous action button improvement, please put their respective commits in two different branches.

MrPetovan commented on view/theme/frio/templates/comment_item.tpl:
```diff
@@ -61,6 +61,49 @@
+  Dropzone.autoDiscover = false;
+  var dropzone{{$id}} = new Dropzone( '#comment-edit-wrapper-{{$id}}', {
+    paramName: "userfile", // The name that will be used to transfer the file
+    maxFilesize: 6, // MB
+    previewsContainer: '#dz-preview-{{$id}}',
+    preventDuplicates: true,
+    clickable: true,
+    thumbnailWidth: 100,
+    thumbnailHeight: 100,
+    url: "/media/photo/upload?response=url&album=",
+    accept: function(file, done) {
+      done();
+    },
+    init: function() {
+      this.on("success", function(file, serverResponse) {
+        var target = $('#comment-edit-text-{{$id}}')
+        var resp = $(serverResponse).find('div#content').text()
+        if (target.setRangeText) {
+          //if setRangeText function is supported by current browser
+          target.setRangeText(" " + $.trim(resp) + " ")
+        } else {
+          target.focus()
+          document.execCommand('insertText', false /*no UI*/, " " + $.trim(resp) + " ");
+        }
+      });
+    },
+  });
+
+  $('#comment-edit-wrapper-{{$id}}').on('paste', function(event){
+    const items = (event.clipboardData || event.originalEvent.clipboardData).items;
+    items.forEach((item) => {
+      if (item.kind === 'file') {
+        // adds the file to your dropzone instance
+        console.log(item);
+        dropzone{{$id}}.addFile(item.getAsFile())
+      }
+    })
+  });
+
+
```
Please use tabulation for indentation.

MrPetovan commented on view/theme/frio/templates/head.tpl:
```diff
@@ -55,6 +55,8 @@
 	media="screen" />
 <link rel="stylesheet" href="view/theme/frio/css/font-awesome.custom.css?v={{$smarty.const.FRIENDICA_VERSION}}"
 	type="text/css" media="screen" />
+<link rel="stylesheet" href="/vendor/enyo/dropzone/dist/min/dropzone.min.css?v={{$smarty.const.FRIENDICA_VERSION}}"
```
```suggestion
<link rel="stylesheet" href="vendor/enyo/dropzone/dist/min/dropzone.min.css?v={{$smarty.const.FRIENDICA_VERSION}}"
```

xundeenergie replied:
removed all occurances of dropzone.min.[js|css] except in view/theme/frio/templates/head.tpl, because this is enough for functionality in all parts.

xundeenergie replied:
This also removed the Browser error that Dropzone already exists in /compose

MrPetovan commented on composer.json:
```diff
@@ -72,7 +72,8 @@
 		"kornrunner/blurhash": "^1.2",
-		"psr/clock": "^1.0"
+		"psr/clock": "^1.0",
+		"enyo/dropzone": "^5.9"
```
Please move this alphabetically after the `"divineomega/password_exposed": "^2.8",` line

MrPetovan reviewed (changes requested):
Thank you for your work! When you add a Composer dependency, you also have to update the list of pinned versions for each dependency. To be as little disruptive as possible, please make sure the only change to `composer.lock` is related to the library you added.

Typically, you should revert the `composer.json` and `composer.lock` file to the `develop` branch version, and run the following command: `bin.composer.phar require enyo/dropzone`. This will update both files the way we want you to commit them.

As for the PR itself, I have a couple general remarks:
- Please use tabulations for indentation
- Please remove commented out code and debug statements before submitting your PR for review.

MrPetovan commented:
I also converted your PR to draft as it is a Work-In-Progress PR. This signals that we shouldn't merge it to the Friendica repository just yet until you work out the last kinks.

xundeenergie commented:
> bin.composer.phar require enyo/dropzone
When i run the command `bin/composer.phar require enyo/dropzone` i get this:
![grafik](https://user-images.githubusercontent.com/5894829/224659858-5f6443e8-f3e9-4174-b5b9-33dfa270c4e5.png)

you surely know this :)

Commit:
fix a warning and suggestions for pullrequest

* fix indentation to tabs
* remove unneeded calls for dropzone.min.js and dropzone.min.css
* removed debug output

Commit:
take maxFilesize from systemsettings

* does not work everywhere...

Commit:
small modifications to equalize the scripts

MrPetovan commented:
For Composer, please disregard Warning and Deprecation Notices stemming from running Composer 1 with PHP 8. Also I see you did everything right committing the `composer.lock` file!

MrPetovan commented on composer.json:
```diff
@@ -29,6 +29,7 @@
 		"asika/simple-console": "^1.0",
 		"bacon/bacon-qr-code": "^2.0.0",
 		"divineomega/password_exposed": "^2.8",
+                "enyo/dropzone": "^5.9",
```
Please follow the indentation convention for JSON files with tabulations.
```suggestion
		"enyo/dropzone": "^5.9",
```

xundeenergie replied:
sorry... my vimrc gets better and better :)

MrPetovan commented on mod/photos.php:
```diff
@@ -1139,7 +1139,11 @@ function photos_content(App $a)
+						'$rand_num' => Crypto::randomDigits(12),
+						// Dropzone
+						//'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
+						// don't know, if DI::config.. does not work here, so it is set to a manual value
```
`DI::config()` should work here.

xundeenergie replied:
it should... but when use it, i get http 500...

MrPetovan replied:
These should produce an error message in your Friendica log that you should enable when you work on it. Notice level should be enough.

xundeenergie replied:
I also changed it to DI, becaus ist listed in the use-headers.

MrPetovan commented on src/Content/Conversation.php:
```diff
@@ -408,6 +408,11 @@ public function statusEditor(array $x = [], int $notes_cid = 0, bool $popup = fa
 			'$always_open_compose' => $this->pConfig->get($this->session->getLocalUserId(), 'frio', 'always_open_compose', false),
+			// Dropzone
+			//'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
+			// DI::config.. does not work here, so it is set to a manual value
```
In this file you should use `$this->config->get('system', 'maximagesize')`, but `DI::config()->get('system', 'maximagesize')` should also work.

MrPetovan commented on src/Module/Post/Edit.php:
```diff
@@ -182,6 +182,11 @@ protected function content(array $request = []): string
 			'$compose_link_title' => $this->t('Open Compose page'),
+			// Dropzone
+			//'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
+			// DI::config.. does not work here, so it is set to a manual value
```
`DI::config()->get('system', 'maximagesize')` should work here.

xundeenergie replied:
I don't know why... but when i use it, i geht a http 500 when i try to edit posting or comment.

xundeenergie replied:
Sorry... i tried both. DI:: and $this... on both i get 500 internal server error, when i try to edit a posting.
with hardcoded Bytesize, it works.

`'$max_imagesize'      => $this->config->get('system', 'maximagesize'),`
`'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),`

xundeenergie replied:
"Not working" means... the edit-overlay does not appear. In Devtools i see the Request with answer "500 Internal Server Error"

Commit:
fix indentation

Commit:
fix missing DI

Commit:
remove console.log

MrPetovan commented on mod/photos.php:
```diff
@@ -1141,9 +1141,7 @@ function photos_content(App $a)
-						//'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
-						// don't know, if DI::config.. does not work here, so it is set to a manual value
-						'$max_imagesize' => 100000,
+						'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
```
So, going back to my initial request, you need the call to `Strings::getBytesFromShortHand` because the `system.maximagesize` value accepts both integer numbers of bytes and string shorthands like `2M` or `2G`. Unless you know that the Dropzone library accepts both, you have to convert the config value to always be an integer, and this is what `Strings::getBytesFromShortHand` does.

xundeenergie replied:
I understand.

Will change this.

xundeenergie replied:
> So, going back to my initial request, you need the call to `Strings::getBytesFromShortHand` because the `system.maximagesize` value accepts both integer numbers of bytes and string shorthands like `2M` or `2G`. Unless you know that the Dropzone library accepts both, you have to convert the config value to always be an integer, and this is what `Strings::getBytesFromShortHand` does.

Again... this is not working...

I have both tried:
```
~                         //'$max_imagesize'       => Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'));
+                         '$max_imagesize'       => \Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'));
```

with and without `use Friendica\Util\Strings;` in the header.

It produces this:

`2023-03-13T19:58:42Z index [CRITICAL]: Uncaught Exception ParseError: "syntax error, unexpected token ";", expecting "]"" at /var/www/soc.schuerz.at/src/Module/Post/Edit.php line 189 {"exception":"ParseError: syntax error, unexpected token \";\", expecting \"]\" in /var/www/soc.schuerz.at/src/Module/Post/Edit.php:189\nStack trace:\n#0 /var/www/soc.schuerz.at/vendor/composer/ClassLoader.php(322): Composer\\Autoload\\includeFile()\n#1 [internal function]: Composer\\Autoload\\ClassLoader->loadClass()\n#2 /var/www/soc.schuerz.at/vendor/level-2/dice/Dice.php(73): is_subclass_of()\n#3 /var/www/soc.schuerz.at/vendor/level-2/dice/Dice.php(93): Dice\\Dice->getRule()\n#4 /var/www/soc.schuerz.at/src/App/Router.php(353): Dice\\Dice->create()\n#5 /var/www/soc.schuerz.at/src/App.php(691): Friendica\\App\\Router->getModule()\n#6 /var/www/soc.schuerz.at/index.php(52): Friendica\\App->runFrontend()\n#7 {main}"} - {"file":null,"line":null,"function":null,"request-id":"640f8072c469e","uid":"51f411","process_id":1267833}
`

xundeenergie replied:
I tried the following:
```
Strings::getBytesFromShorthand('5G');
Strings::getBytesFromShorthand(5000000);
```
Both gave me the same error, that the line expects "]" instead of ";"

xundeenergie replied:
In src/Model/Photo.php is the same call... and obvious it works there.

Someshing ist missing?

MrPetovan replied:
Yes, template variables are declared inside of an array which values should be separated by commas, not semi-colon. The snippet I provided to you only works if it's on its own line, not when it is used inside an array declaration.

Alongside my other change request:
```suggestion
						'$max_imagesize'       => floor(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000),
```

MrPetovan commented on view/templates/item/compose.tpl:
```diff
@@ -97,7 +97,6 @@
 <script>
 	Dropzone.autoDiscover = false;
-	console.log('compose.tpl', {{$max_imagesize}} / 100000);
 	var maxis = {{$max_imagesize}} / 100000;
```
This doesn't look correct, the divisor probably should be `1000000` (10^6). This said, I would prefer if this calculation was performed when the template value is set rather than here.

tobiasd commented:
If you feel that this is stable enough, I think it would be worth on changing the targeted branch from `develop` to `2023.03-rc`.

xundeenergie commented:
> If you feel that this is stable enough, I think it would be worth on changing the targeted branch from `develop` to `2023.03-rc`.

I have one change, i'm testing.
Because C&P and D&D did not work on editing a comment or posting.
Now it works, but the upload-Button is not shown, and i'm not sure, why. But, it works in Browser.

I'll make the push in a few minutes. Have to clean up debugging-Output.

tobiasd commented:
Dont forget, there are then still 2 weeks left fixing bugs in this new feature ;-) I really think this is a cool new feature to add last minute to the RC branch.

Commit:
make dropzone work on editing postings/Comments

Commit:
create workaround for not working php

Strings:getBytesFromShorthand

i created a js-function which converts integer, k, m and g to MB
Just remove this code, when we found out, why Strings is in some files
not working.

Commit:
remove unneeded chars ("css") from first html-tag

xundeenergie commented:
> > If you feel that this is stable enough, I think it would be worth on changing the targeted branch from `develop` to `2023.03-rc`.
> 
> I have one change, i'm testing. Because C&P and D&D did not work on editing a comment or posting. Now it works, but the upload-Button is not shown, and i'm not sure, why. But, it works in Browser.
> 
> I'll make the push in a few minutes. Have to clean up debugging-Output.

Thanks :)

Yes this was really last minute... 

It's my first time, working on such a project. What's todo for me now? Should i also rebase to 2023.03-rc? Or whats todo for me locally?

MrPetovan commented:
Ideally, yes, although since the RC branches forks off develop, it shouldn’t be too involved. I’ve changed the merge target branch myself.

Commit:
improve image-upload

* create dropzone for editing in jot-modal
* customize css for dropzone to make it more beautiful (smaller
  margin/paddings...)
* change position of dropzone-element dircetly to textarea, so button
  for upload is small and under textarea (it's more clear)
* remove previews after upload. this shows the upload-progress, but
  preview is not needed after upload. you can't do anything with the
  preview, and dialog gets too big. You see successful upload, when
  bb-code link to image is inserted on cursor-position in textarea

xundeenergie commented:
Puh... i hope this rebasing was successful...

i did `git rebase 2023.03-rc` on branch easy-image-attach

but when i look in git log, i see this

![grafik](https://user-images.githubusercontent.com/5894829/224975783-8270973b-f135-4bde-afe1-6e688e7db2b0.png)

seems wrong. Isn't it?

xundeenergie commented:
> Puh... i hope this rebasing was successful...
> 
> i did `git rebase 2023.03-rc` on branch easy-image-attach
> 
> but when i look in git log, i see this
> 
> ![grafik](https://user-images.githubusercontent.com/5894829/224975783-8270973b-f135-4bde-afe1-6e688e7db2b0.png)
> 
> seems wrong. Isn't it?

Hmmm... when i rebase to 2023.03-rc i can not push anymore to github (my repo).
i have to pull and resolve mergeconflicts.
Then this branch is again not based anymore on 2023.03-rc... i have to rebase again.

How do i have to do this?

xundeenergie commented:
Puh... rebasing... i made a mistake... but i could fix it hopefully.
now my branch is rebased to 2023.03-rc.

Commit:
add timeout before remove images from dropzone

Commit:
fix editing in modal-jot

and add also timeout before removing images from dropzone

MrPetovan commented on mod/photos.php:
```diff
@@ -1194,7 +1198,11 @@ function photos_content(App $a)
+						// Dropzone
+                                                //'$max_imagesize'       => Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'));
+                                                // Strings::getBytresFromShorthand not working, so just workaround in template and serve the bare value
+						'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
```
```suggestion
						'$max_imagesize' => floor(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000),
```

MrPetovan commented on mod/photos.php:
```diff
@@ -1268,7 +1276,11 @@ function photos_content(App $a)
+							// Dropzone
+                                                        //'$max_imagesize'       => Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'));
+                                                        // Strings::getBytresFromShorthand not working, so just workaround in template and serve the bare value
+							'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
```
```suggestion
							'$max_imagesize' => floor(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000),
```

MrPetovan commented on src/Content/Conversation.php:
```diff
@@ -408,6 +409,12 @@ public function statusEditor(array $x = [], int $notes_cid = 0, bool $popup = fa
 			'$always_open_compose' => $this->pConfig->get($this->session->getLocalUserId(), 'frio', 'always_open_compose', false),
+
+			// Dropzone
+                        //'$max_imagesize'       => Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'));
+                        // Strings::getBytresFromShorthand not working, so just workaround in template and serve the bare value
+			'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
+
```
```suggestion
			'$max_imagesize' => floor(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000),
```

MrPetovan commented on mod/photos.php:
```diff
@@ -1139,7 +1139,11 @@ function photos_content(App $a)
-						'$rand_num' => Crypto::randomDigits(12)
+						'$rand_num' => Crypto::randomDigits(12),
+						// Dropzone
+                                                //'$max_imagesize'       => Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'));
+                                                // Strings::getBytresFromShorthand not working, so just workaround in template and serve the bare value
+						'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
```
```suggestion
						'$max_imagesize' => floor(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000),
```

MrPetovan commented on src/Module/Item/Compose.php:
```diff
@@ -241,6 +240,9 @@ protected function content(array $request = []): string
 			]),
+
+			// Dropzone
+			'$max_imagesize'    => DI::config()->get('system', 'maximagesize')
```
```suggestion
			'$max_imagesize' => floor(\Friendica\Util\Strings::getBytesFromShorthand($this->config->get('system', 'maximagesize')) / 1000000),

```

MrPetovan commented on src/Module/Post/Edit.php:
```diff
@@ -182,6 +184,11 @@ protected function content(array $request = []): string
 
 			'$compose_link_title' => $this->t('Open Compose page'),
+			// Dropzone
+			//'$max_imagesize'       => \Friendica\\Util\\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'));
+			//'$max_imagesize'       => Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize'));
+                        // both not working, so just workaround in template and serve the bare value
+			'$max_imagesize'       => DI::config()->get('system', 'maximagesize'),
```
```suggestion
			'$max_imagesize' => floor(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000),
```

MrPetovan commented on view/theme/frio/js/dropzone-frio.js:
```diff
@@ -0,0 +1,24 @@
+	return /^\d+$/.test(value);
+}
+
+function getMBytes(value) {
```
This works but I'd rather we didn't have the same method duplicated in both PHP and Javascript. Please perform the calculation in the PHP module like I suggested and remove this file.

Commit:
fix getting bytes from shorthand in php

xundeenergie commented:
I changed it. Thanks for your help.

BUT

The code gives MB back. No division by 10^6 necessary. 
And, it only works for m and g and numbers >= 1000000 and only integer.

Should i file an issue? because, if someone sets to 500k or 999999 the returned value is 0 and no limit is set... it falls on the bad side.

MrPetovan commented:
What is returning a number of MB exactly?

xundeenergie commented:
> Puh... rebasing... i made a mistake... but i could fix it hopefully. now my branch is rebased to 2023.03-rc.

@MrPetovan one question on rebasing... 

When i rebase to another branch (2023.03-rc for example), i have to push --force to github again to my repo... but forced push is not recommended... 

How do i do it right?

xundeenergie commented:
> What is returning a number of MB exactly?

An Integer.

When i set 5m i get 5 back.
When i set 5g, i get 5000 (or a little more, must look later)
When i set 5k, i get 0 back.

The same with 5000000 i get 5 back.
With 999999 i get 0 back.

MrPetovan commented:
What method/function are you talking about?

xundeenergie commented:
> 

This one
`floor(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000)`

I changed to round with precision of 1. https://www.php.net/manual/de/function.round.php
Which does a good job.

xundeenergie commented:
> > 
> 
> This one `floor(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000)`


Ah, for sure... floor gives the integer... 😄😄😄

Commit:
create factory for dropzones

Commit:
create factory

Commit:
attach dropzone only when opening comment

xundeenergie commented:
With my last commit i reworked the creation of dropzone...

I added a factory and on comments dropzone for a comment-filed will only be created, when user click on comment-button. 
It should use much less memory and load site faster.

The code for creating dropzones is now on only one place to configure in `view/js/dropzone-factory.js`

MrPetovan commented:
> Ah, for sure... floor gives the integer... 😄😄😄

Indeed. I believe you need it if the parameter expects an integer number of MB, and you can set a minimum value of 1 like this:
```
max(1, floor(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000))
```

Commit:
use round instead of floor to make imagesize <1M possible

MrPetovan commented on mod/photos.php:
```diff
@@ -1139,7 +1139,9 @@ function photos_content(App $a)
-						'$rand_num' => Crypto::randomDigits(12)
+						'$rand_num' => Crypto::randomDigits(12),
+						// Dropzone
+						'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000,1),
```
Code standards: Please add a space after commas.
```suggestion
						'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000, 1),
```

MrPetovan commented on mod/photos.php:
```diff
@@ -1194,7 +1196,9 @@ function photos_content(App $a)
-						'$rand_num' => Crypto::randomDigits(12)
+						'$rand_num' => Crypto::randomDigits(12),
+						// Dropzone
+						'$max_imagesize' => round(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000,1),
```
Code standards: Please add a space after commas.
```suggestion
						'$max_imagesize' => round(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000, 1),
```

MrPetovan commented on mod/photos.php:
```diff
@@ -1268,7 +1272,9 @@ function photos_content(App $a)
-							'$rand_num' => Crypto::randomDigits(12)
+							'$rand_num' => Crypto::randomDigits(12),
+							// Dropzone
+							'$max_imagesize' => round(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000,1),
```
Code standards: Please add a space after commas.
```suggestion
							'$max_imagesize' => round(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000, 1),
```

MrPetovan commented on src/Content/Conversation.php:
```diff
@@ -408,6 +409,10 @@ public function statusEditor(array $x = [], int $notes_cid = 0, bool $popup = fa
 			'$always_open_compose' => $this->pConfig->get($this->session->getLocalUserId(), 'frio', 'always_open_compose', false),
+
+			// Dropzone
+			'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000,1),
```
Code standards: Please add a space after commas.

Additionally, a local property provides access to the config
```suggestion
			'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand($this->config->get('system', 'maximagesize')) / 1000000, 1),
```

MrPetovan commented on src/Content/Conversation.php:
```diff
@@ -35,6 +35,7 @@
 use Friendica\Core\Session\Capability\IHandleUserSessions;
 use Friendica\Core\Theme;
 use Friendica\Database\DBA;
+use Friendica\DI;
```
No need after using the local `config` property.
```suggestion
```

MrPetovan commented on src/Module/Item/Compose.php:
```diff
@@ -241,6 +240,9 @@ protected function content(array $request = []): string
 			]),
+
+			// Dropzone
+			'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000,1),
```
Code standards: Please add a space after commas.

Additionally, you can use the local `config` variable.
```suggestion
			'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand($this->config->get('system', 'maximagesize')) / 1000000, 1),
```

MrPetovan commented on src/Module/Post/Edit.php:
```diff
@@ -182,8 +184,10 @@ protected function content(array $request = []): string
-		]);
 
+			// Dropzone
+			'$max_imagesize'       => round(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000,1),
```
Code standards: Please add a space after commas.
```suggestion
			'$max_imagesize'       => round(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000, 1),
```

MrPetovan commented on src/Object/Post.php:
```diff
@@ -1068,7 +1068,9 @@ private function getCommentBox(string $indent)
-				'$rand_num'    => Crypto::randomDigits(12)
+				'$rand_num'    => Crypto::randomDigits(12),
+				// Dropzone
+				'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000,1),
```
Code standards: Please add a space after commas.
```suggestion
				'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000, 1),
```

Commit:
add dropzone also onclick on comment-fake-form

MrPetovan commented on view/theme/frio/templates/jot.tpl:
```diff
@@ -162,19 +165,63 @@
+
+<script>
+	Dropzone.autoDiscover = false;
+	var dropzoneJot = new Dropzone( '#jot-text-wrap', {
```
Is this direct instantiation still needed now that you created the DropzoneFactory?

xundeenergie replied:
for sure not.
Just forgot to change this and in view/templates/item/compose.tpl too, to the new method.

Commit:
make errormessages visible

Commit:
apply changes from review

xundeenergie commented:
> Indeed. I believe you need it if the parameter expects an integer number of MB, and you can set a minimum value of 1 like this:
> 
> ```
> max(1, floor(Strings::getBytesFromShorthand(DI::config()->get('system', 'maximagesize')) / 1000000))
> ```

xundeenergie commented:
Sorry... i deleted the wrong remote... i pushed it again... should create a new pullrequest?

MrPetovan commented:
It's ok, we're back on track.

xundeenergie commented:
> It's ok, we're back on track.

Puh... i have to learn much about git... :D

MrPetovan commented on view/theme/frio/js/dropzone-frio.js:
Please remove this file now that its functions aren't used anymore.

xundeenergie replied:
What's better, to place it here, or in view/js?

I was not sure, if it's better theme-specific, or general useable (so other themes can modified too...

MrPetovan replied:
In this case, there's little doubt since none of its functions are used anywhere.

Beyond this case, it's entirely up to you to choose the scope of your changes. It's more work to make a theme-agnostic feature but now that we are down to only two officially-supported themes, it's less time-consuming than when we had 5.

MrPetovan commented on view/theme/frio/js/modal.js:
```diff
@@ -291,6 +291,51 @@ function editpost(url) {
 
+			// To make dropzone fileupload work on editing a comment, we need to
+			// attach a new dropzone to modal
+			dropzoneJotEdit = new Dropzone( '#jot-text-wrap', {
```
Can this be replaced by a `setupDropzone` call?

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,63 @@
+var DzFactory = function () {
+	this.createDropzone = function(element, target, maxImagesize) {
```
Please rename `element` to `dropSelector` and `target` parameter to `textareaSelector`.
```suggestion
	this.createDropzone = function(dropSelector, textareaSelector, maxImagesize) {
```

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,63 @@
+var DzFactory = function () {
+	this.createDropzone = function(element, target, maxImagesize) {
+		return new Dropzone( element, {
```
Code standards.
```suggestion
		return new Dropzone(dropSelector, {
```

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,63 @@
+			},
+			init: function() {
+				this.on('success', function(file, serverResponse) {
+					var _target = $(target)
+					var resp = $(serverResponse).find('div#content').text()
+					if (_target.setRangeText) {
+						//if setRangeText function is supported by current browser
+						_target.setRangeText(' ' + $.trim(resp) + ' ')
+					} else {
+						_target.focus()
+						document.execCommand('insertText', false /*no UI*/, ' ' + $.trim(resp) + ' ');
+					}
```
A few things here:
- Please avoid the overly-general `var` in favor of `let` (same meaning as PHP) or `const` (if the value isn't re-initialized.
- Please add a semi-colon (;) at the end of each statement even if it isn't necessary.
- Please avoid vague variable names like "target" or "response" in favor of more specific names.
- You can combine the string operations of `bbcodeString` in one place.
- I like to prepend variable names with a dollar sign when it contains a jQuery object. This also allowed me to realize you are trying to call [`HTMLInputElement.setRangeText`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setRangeText) on a jQuery object, which probably won't work. Please review and address.
```suggestion
					const $targetTextarea = $(textareaSelector);
					const bbcodeString = ' ' + $.trim($(serverResponse).find('div#content').text()) + ' ';
					if ($targetTextarea.setRangeText) {
						// if setRangeText function is supported by current browser
						$targetTextarea.setRangeText(bbcodeString);
					} else {
						$targetTextarea.focus();
						document.execCommand('insertText', false /*no UI*/, bbcodeString);
					}
```

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,63 @@
+	this.setupDropzone = function(element, target, maxImagesize) {
+		var dropzone = this.createDropzone(element, target, maxImagesize)
+		$(element).on('paste', function(event) {
+
```
```suggestion
```

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,63 @@
+		})
+	};
+
+	this.setupDropzone = function(element, target, maxImagesize) {
+		var dropzone = this.createDropzone(element, target, maxImagesize)
```
```suggestion
	this.setupDropzone = function(dropSelector, textareaSelector, maxImagesize) {
		var dropzone = this.createDropzone(dropSelector, textareaSelector, maxImagesize)
```

xundeenergie replied:
this should be aldo done in create Dropzone? Replace target with textareaSelector?
or does this lead to problems with namespace?

MrPetovan commented on view/theme/frio/templates/comment_item.tpl:
```diff
@@ -64,3 +65,15 @@
 
+<script>
+	$('[id=comment-fake-text-{{$id}}]').on('focus', function() {
+		dzFactory.setupDropzone('#dropzone-{{$id}}', $('#comment-edit-text-{{$id}}'), {{$max_imagesize}}); 
```
The jQuery call is unnecessary here since it's done again in `view/js/dropzone-factory.js:12`.
```suggestion
		dzFactory.setupDropzone('#dropzone-{{$id}}', '#comment-edit-text-{{$id}}', {{$max_imagesize}}); 
```

MrPetovan commented on view/theme/frio/templates/comment_item.tpl:
```diff
@@ -64,3 +65,15 @@
+		$('[id=comment-{{$id}}]') .prop('click', null).off('click')
+	});
+	$('[id=comment-{{$id}}]').on('click', function() {
+		dzFactory.setupDropzone('#dropzone-{{$id}}', $('#comment-edit-text-{{$id}}'), {{$max_imagesize}}); 
```
```suggestion
		dzFactory.setupDropzone('#dropzone-{{$id}}', '#comment-edit-text-{{$id}}', {{$max_imagesize}}); 
```

MrPetovan commented on view/theme/frio/templates/comment_item.tpl:
```diff
@@ -64,3 +65,15 @@
+<script>
+	$('[id=comment-fake-text-{{$id}}]').on('focus', function() {
+		dzFactory.setupDropzone('#dropzone-{{$id}}', $('#comment-edit-text-{{$id}}'), {{$max_imagesize}}); 
+		$('[id=comment-fake-text-{{$id}}]') .prop('focus', null).off('focus')
+		$('[id=comment-{{$id}}]') .prop('click', null).off('click')
```
```suggestion
		$('[id=comment-fake-text-{{$id}}]').prop('focus', null).off('focus')
		$('[id=comment-{{$id}}]').prop('click', null).off('click')
```

MrPetovan commented on view/theme/frio/templates/comment_item.tpl:
```diff
@@ -64,3 +65,15 @@
+	});
+	$('[id=comment-{{$id}}]').on('click', function() {
+		dzFactory.setupDropzone('#dropzone-{{$id}}', $('#comment-edit-text-{{$id}}'), {{$max_imagesize}}); 
+		$('[id=comment-fake-text-{{$id}}]') .prop('focus', null).off('focus')
+		$('[id=comment-{{$id}}]') .prop('click', null).off('click')
```
```suggestion
		$('[id=comment-fake-text-{{$id}}]').prop('focus', null).off('focus')
		$('[id=comment-{{$id}}]').prop('click', null).off('click')
```

MrPetovan commented on view/theme/frio/templates/jot.tpl:
```diff
@@ -162,19 +163,22 @@
 </script>
+
+<script>
+	dzFactory.setupDropzone('#jot-text-wrap', $('#profile-jot-text'), {{$max_imagesize}}); 
```
```suggestion
	dzFactory.setupDropzone('#jot-text-wrap', '#profile-jot-text', {{$max_imagesize}}); 
```

xundeenergie replied:
Removed this also in 
- view/templates/item/compose.tpl
- view/theme/frio/js/modal.js

Commit:
changes in case of codereview

xundeenergie commented:
I have one more question, i just realised... 

Here in view/theme/frio/js/modal.js 
https://github.com/xundeenergie/friendica/blob/c72668285a8ed7853bed09a1ff2a0f965dbe2fdf/view/theme/frio/js/modal.js#L297

i need the configuration-variable $max_imagesize, but this is not rendered by smarty3.

How can i get this variable?
Could it be a better idea, to set it in head.tpl, where i also create the factory? So it is available on the whole page?
Or is there another way, to reach this config-variable from javascript?

Commit:
render max_imagesize in header.tpl

so it is reachable general from the whole website, and the dropzone-factory can use
it also to create dropzones from modal.js

xundeenergie commented:
> I have one more question, i just realised...
> 
> Here in view/theme/frio/js/modal.js https://github.com/xundeenergie/friendica/blob/c72668285a8ed7853bed09a1ff2a0f965dbe2fdf/view/theme/frio/js/modal.js#L297
> 
> i need the configuration-variable $max_imagesize, but this is not rendered by smarty3.
> 
> How can i get this variable? Could it be a better idea, to set it in head.tpl, where i also create the factory? So it is available on the whole page? Or is there another way, to reach this config-variable from javascript?

i do the rendering now in head.tpl (from frio), and removed all the renderings of max_imagesize in the 5 php-files from before.
now it works also for modal.js (because it is not needed there anymore)

Commit:
reinsert Friendica\DI (accidentially removed before)

Commit:
add options allowedFiles and more

MrPetovan commented on src/Module/Post/Edit.php:
None of the changes in this file are needed anymore.

xundeenergie replied:
i reverted this file to 2023.03-rc

MrPetovan replied:
The PR diff still shows changes in this file.

MrPetovan commented on src/App/Page.php:
```diff
@@ -253,11 +253,15 @@ private function initHead(App $app, Arguments $args, L10n $l10n, IManageConfigVa
+			'$netError'        => $l10n->t('Network error'),
+			// Dropzone
+			'$max_imagesize' => round(\Friendica\Util\Strings::getBytesFromShorthand($config->get('system', 'maximagesize')) / 1000000, 1),
+
+
```
Code standards: Please remove extra new lines.
```suggestion
```

MrPetovan replied:
This hasn't been addressed.

MrPetovan commented on src/Content/Conversation.php:
```diff
@@ -408,6 +408,7 @@ public function statusEditor(array $x = [], int $notes_cid = 0, bool $popup = fa
 
 			'$compose_link_title'  => $this->l10n->t('Open Compose page'),
 			'$always_open_compose' => $this->pConfig->get($this->session->getLocalUserId(), 'frio', 'always_open_compose', false),
+
```
```suggestion
```

MrPetovan commented on src/Module/Item/Compose.php:
The change in this file isn't needed anymore.

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,65 @@
+			},
+			init: function() {
+				this.on('success', function(file, serverResponse) {
+					const targetTextarea = document.getElementById(textareaSelector);
```
I see you have changed the jQuery call to `getElementById`, so the parameter name should change as well
```suggestion
	this.createDropzone = function(dropSelector, textareaElementId) {
		return new Dropzone(dropSelector, {
			paramName: 'userfile', // The name that will be used to transfer the file
			maxFilesize: max_imagesize, // MB
			url: '/media/photo/upload?album=',
			addRemoveLinks: true,
			acceptedFiles: 'image/*',
			clickable: true,
			accept: function(file, done) {
				done();
			},
			init: function() {
				this.on('success', function(file, serverResponse) {
					const targetTextarea = document.getElementById(textareaElementId);
```

xundeenergie replied:
yes. this was necessary to make setRangeText working as you suggested, that this won't work on jQuery-objects.

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,65 @@
+			init: function() {
+				this.on('success', function(file, serverResponse) {
+					const targetTextarea = document.getElementById(textareaSelector);
+					const bbcodeString = $(serverResponse).find('div#content').text();
```
Once #12892 is merged, this will have to be changed.

xundeenergie replied:
Ok.
Also the request-url i think :)

MrPetovan replied:
Not anymore, #12892 will take care of that.

xundeenergie replied:
Ah, i understand... it was a bug to show the whole page.

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,65 @@
+					}
+				});
+				this.on('complete', function(file) {
+					var dz = this;
```
```suggestion
					const dz = this;
```

MrPetovan commented on view/theme/frio/templates/comment_item.tpl:
```diff
@@ -64,3 +65,15 @@
+	$('[id=comment-{{$id}}]').on('click', function() {
+		dzFactory.setupDropzone('#dropzone-{{$id}}', 'comment-edit-text-{{$id}}'); 
+		$('[id=comment-fake-text-{{$id}}]').prop('focus', null).off('focus')
+		$('[id=comment-{{$id}}]').prop('click', null).off('click')
```
```suggestion
		$('[id=comment-fake-text-{{$id}}]').prop('focus', null).off('focus');
		$('[id=comment-{{$id}}]').prop('click', null).off('click');
	});
	$('[id=comment-{{$id}}]').on('click', function() {
		dzFactory.setupDropzone('#dropzone-{{$id}}', 'comment-edit-text-{{$id}}'); 
		$('[id=comment-fake-text-{{$id}}]').prop('focus', null).off('focus');
		$('[id=comment-{{$id}}]').prop('click', null).off('click');
```

MrPetovan commented on view/theme/frio/templates/head.tpl:
```diff
@@ -137,6 +139,9 @@
 	<script type="text/javascript" src="view/theme/frio/js/textedit.js?v={{$smarty.const.FRIENDICA_VERSION}}"></script>
+	<script type="text/javascript" src="vendor/enyo/dropzone/dist/min/dropzone.min.js?v={{$smarty.const.FRIENDICA_VERSION}}"></script>
+	<script type="text/javascript" src="view/js/dropzone-factory.js?v={{$smarty.const.FRIENDICA_VERSION}}"></script>
+	<script type="text/javascript"> max_imagesize = {{$max_imagesize}}; var dzFactory = new DzFactory(); Dropzone.autoDiscover = false; </script>
```
```suggestion
	<script type="text/javascript">
		var max_imagesize = {{$max_imagesize}};
		var dzFactory = new DzFactory();
		Dropzone.autoDiscover = false;
	</script>
```

Commit:
changes in case of code-review

Commit:
remove upload-url parameter

Commit:
revert to original (no changes needed anymore)

MrPetovan commented on src/Object/Post.php:
Please revert the changes in this file as well.

xundeenergie replied:
strange... i did them all before... now it should be fixed.

Commit:
revert changes, remove line (from code-review)

MrPetovan commented:
You did it, you got an approval! Since this PR is still in Draft, i won't merge it until you decide it's ready to be merged. In this case, it mainly hinges on https://github.com/friendica/friendica/pull/12892 being merged since it will affect your code.

Now, if you could test my changes with this PR, it would be really good (for me, that is).

xundeenergie commented:
I tested your pullrequest https://github.com/friendica/friendica/pull/12892 against my code, changed handling of the response... and it workes!!

So i will push this change again, and i'm fine for merging it.
I think, it does, what it should :)

Commit:
change responsehandling to be ready for PR #12892

xundeenergie commented:
So if you want, you can merge it together with your PR https://github.com/friendica/friendica/pull/12892

Should i change the draft-status, or do you have to do that?

MrPetovan commented on view/js/dropzone-factory.js:
```diff
@@ -0,0 +1,65 @@
+			init: function() {
+				this.on('success', function(file, serverResponse) {
+					const targetTextarea = document.getElementById(textareaElementId);
+					const bbcodeString = serverResponse;
+					if (targetTextarea.setRangeText) {
+						//if setRangeText function is supported by current browser
+						targetTextarea.setRangeText(' ' + $.trim(bbcodeString) + ' ');
+					} else {
+						targetTextarea.focus();
+						document.execCommand('insertText', false /*no UI*/, '\n' + $.trim(bbcodeString) + '\n');
```
```suggestion
					if (targetTextarea.setRangeText) {
						//if setRangeText function is supported by current browser
						targetTextarea.setRangeText(' ' + $.trim(serverResponse) + ' ');
					} else {
						targetTextarea.focus();
						document.execCommand('insertText', false /*no UI*/, '\n' + $.trim(serverResponse) + '\n');
```

xundeenergie replied:
Changed it.

MrPetovan replied:
Thanks, I'm still waiting on #12892 to be merged since this PR is now depending on it.

Commit:
change response-handling (code-review)

MrPetovan commented:
Now depends on #12892

xundeenergie commented:
Is it ok, if i enhance it for all attachments now?
It's a small change to do the upload for images to `/media/photo/upload` and for all other files to `/media/attachment/upload`

Or should i wait for https://github.com/friendica/friendica/issues/12902 ?

MrPetovan commented:
Let's wait first for https://github.com/friendica/friendica/pull/12892, then this PR, and then https://github.com/friendica/friendica/issues/12902 to be merged/resolved.

Of course you're welcome to look into https://github.com/friendica/friendica/issues/12902 yourself in parallel.

MrPetovan commented:
Now that #12892 has been merged, can you please rebase this branch onto the latest RC branch and confirm it still works as intended?

Commit:
remove removelink (it is not useful at the moment)

xundeenergie commented:
I've tested successfully all cases after rebasing to the latest rc commit.

And i removed the "removebutton" option, because it is misleading and nonfunctional at the moment.
Will implement it later.

xundeenergie commented:
One question... 
github shows me one open change request.. but this is resolved, when i got to it... 
![grafik](https://user-images.githubusercontent.com/5894829/226198900-263dd4a9-8ee1-40f0-98fd-b79c264c57fb.png)

What does this mean?

nupplaphil commented:
@MrPetovan requested changes and he doesn't "accept" your current changes yet, that's why this message appears.
For me, the PR looks good now so I "approved" it.

That's why the circle is half green, half red :)

We have to wait for @MrPetovan and he can now decide to merge the PR

xundeenergie commented:
Ah ok.
Thank you.

xundeenergie commented:
Thank you for your help!

MrPetovan commented:
Thank you for your work! You're about to become a kind of folk hero in the tiny Friendica community, thanks to your frontend work!

xundeenergie commented:
😊

I just wanted to have this features... 😇
